### PR TITLE
Implement optional dense indexes

### PR DESCRIPTION
**Pull request by patszt**
Creation and usage of optional dense index can be optional for a given type or field name.
It can be configured in vast.yaml under index in each rule.
By default the index is always created.

### :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [vast.io](https://vast.io), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Starting with docs might be useful to grasp what the change is supposed to bring

### CODE REVIEW

File: changelog/unreleased/features/2430--optional-dense-indexes.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+VAST can be configured to not create dense indexes for specific fields or
+types. To disable the creation of dense indexes the user should modify
+the index section in .yaml configuration file. For more informations see
+[here](https://vast.io/docs/setup-vast/tune#skip-dense-index-creation).
```
Comment (dominiklohmann) [inline changelog/unreleased/features/2430--optional-dense-indexes.md:4]: Please reword this to use active voice and present tense.

File: web/docs/setup-vast/tune.md [hunk 2]
```diff
@@ -184,6 +183,34 @@ applies to a single field, `suricata.http.http.url`, and has false-positive rate
...
+        # don't create dense indexes for suricata.http.http.url fields
+      - targets:
+          - suricata.http.http.url
+        create-dense-index: false
+        # don't create dense indexes for addr type columns
+      - targets:
+          - :addr
+        create-dense-index: false
+```
+
```
Comment (dominiklohmann) [inline web/docs/setup-vast/tune.md:213]: Please reword this to use active voice and present tense.

Lead with what problem this solves for the user so that it's clear why they'd want to do this in the first place, i.e., the disk and memory usage vs. query performance trade-off.

**Review by dominiklohmann — COMMENTED** [review]
I only looked at the docs so far; I'll leave the actual code review to the rest of the team.

File: web/docs/setup-vast/tune.md (hunk 2 above)
Comment (patszt) [reply web/docs/setup-vast/tune.md:213]: I am not sure if memory usage is higher with the indexers. It can be, but it also may be not noticeable to the user.
Should i test it separately or just ignore the memory usage information?

File: vast.yaml.example [hunk 3]
```diff
@@ -156,12 +156,17 @@ vast:
...
     # rules:
-    #   Every rule adjusts the false-positive rate of a set of targets.
+    #   Every rule adjusts the behaviour of VAST for a set of targets.
     #   VAST creates one synopsis per target. Targets can be either types
-    #   or field names. All rules for types other than string and address
-    #   are currently ignored.
+    #   or field names.
+    #
+    #   fp-rate - false positive rate. Only string and address type targets
+    #             are supported
```
Comment (lava) [inline vast.yaml.example:164]: I'd probably say will only have an effect for string an address type targets.

File: libvast/include/vast/system/evaluator.hpp [hunk 4]
```diff
@@ -76,6 +83,7 @@ struct evaluator_state {
 /// @pre `!eval.empty()`
 evaluator_actor::behavior_type
 evaluator(evaluator_actor::stateful_pointer<evaluator_state> self,
-          expression expr, std::vector<evaluation_triple> eval);
+          expression expr, std::vector<evaluation_triple> eval,
+          ids all_possible_ids_for_partition);
```
Comment (lava) [inline libvast/include/vast/system/evaluator.hpp:87]: Maybe this should be `optional<ids>` since you sometimes also send an empty `ids{}`

File: libvast/src/detail/partition_common.cpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  });
+}
+} // namespace
+
+ids get_ids_for_evaluation(
+  const std::unordered_map<std::string, ids>& type_ids,
+  const std::vector<system::evaluation_triple>& evaluation_triples) {
+  if (needs_evaluation_without_indexer(evaluation_triples)) {
+    auto all_ids = vast::ids{};
+    for (const auto& [_, ids] : type_ids) {
```
Comment (lava) [inline libvast/src/detail/partition_common.cpp:32]: Current style is to not use braces when the body is just a single line of code.

File: libvast/src/system/active_partition.cpp [hunk 6]
```diff
@@ -62,6 +62,129 @@
...
+  fbs::PartitionSynopsisBuilder ps_builder(synopsis_builder);
+  ps_builder.add_partition_synopsis_type(
+    fbs::partition_synopsis::PartitionSynopsis::legacy);
+  ps_builder.add_partition_synopsis(ps->Union());
+  auto ps_offset = ps_builder.Finish();
+  fbs::FinishPartitionSynopsisBuffer(synopsis_builder, ps_offset);
+  return fbs::release(synopsis_builder);
+}
+
+void serialize(
```
Comment (lava) [inline libvast/src/system/active_partition.cpp:92]: I think this needs a small comment to explain the effect it has on the state.

File: libvast/src/system/active_partition.cpp [hunk 7]
```diff
@@ -412,114 +539,12 @@ active_partition_actor::behavior_type active_partition(
                            *self, self->state.persisted_indexers, valid_count);
                 return;
```
Comment (lava) [inline libvast/src/system/active_partition.cpp:540]: The old code was not pretty here, nowadays we do have `detail::fanout_counter` to handle this pattern. It may make sense to use it here and then call `serialize()` from its completion handler.

**Review by lava — APPROVED** [review]
Looks great, and very good test coverage!

File: libvast/include/vast/system/evaluator.hpp (hunk 4 above)
Comment (patszt) [reply libvast/include/vast/system/evaluator.hpp:87]: I have changed variable name so it no longer "lies"

File: libvast/src/system/active_partition.cpp [hunk 8]
```diff
@@ -62,6 +62,129 @@
...
 namespace vast::system {
 
+namespace {
+
+bool should_skip_index_creation(const type& type,
+                                const qualified_record_field& qf,
+                                const std::vector<index_config::rule>& rules) {
+  if (type.attribute("skip").has_value()) {
+    return true;
+  }
```
Comment (lava) [inline libvast/src/system/active_partition.cpp:72]: I just noticed we're also fixing sc-35515 with this PR.

Can you add one more integration test that imports anything with the `#skip` attribute, flushes the db, and then exports that event again to ensure this works, and add a separate changelog entry (ie. a new paragraph in the `.md` file) for the bugfix?

Comment (patszt) [reply libvast/src/system/active_partition.cpp]: Will be added soon

File: web/docs/setup-vast/tune.md [hunk 9]
```diff
@@ -184,6 +184,32 @@ applies to a single field, `suricata.http.http.url`, and has false-positive rate
...
+  creation rule should be applied.
+- `partition-index` Doesn't create partition indexes for given targets if set to
+  false. By default the value is set to true.
+#### Example
+
+```yaml
+vast:
+  index:
+    rules:
+        # don't create partition indexes for suricata.http.http.url fields
```
Comment (dominiklohmann) [inline web/docs/setup-vast/tune.md:203]:
```suggestion
        # Don't create partition indexes the suricata.http.http.url field.
```

**Review by dominiklohmann — APPROVED** [review]
Thanks for updating the docs. Just some minor requests for improvements.

File: web/docs/setup-vast/tune.md (hunk 2 above)
Comment (dominiklohmann) [reply web/docs/setup-vast/tune.md]: I think we can omit it for now, the disk usage is much more important.
